### PR TITLE
cupy + arrow batching for low_rank_matrix

### PR DESCRIPTION
**Pull request by rishic3**
- Low rank matrix changes:
   * Replaced numpy array operations with cupy
   * Added loop in UDF to yield data in maxRecordsPerBatch chunks for memory efficiency

#### speedups:

3000 cols, float32

- Low rank matrix:

| Samples | Numpy Runtime | Cupy Runtime |
| -------- | -------- | -------- |
| 100,000  | 83s | 43s   |
| 1,000,000  | 603s  | 335s   |

- Regression (low-rank case):

| Samples | Numpy Runtime | Cupy Runtime |
| -------- | -------- | -------- |
| 100,000  | 80s | 67s   |
| 1,000,000  | 675s  | 366s |

If these changes look OK, I'll add arrow batching to the rest of the functions, as well as cupy operations in the regression udf.

### CODE REVIEW

**Comment by rishic3** [comment]
build

**Comment by rishic3** [comment]
build

**Comment by wbo4958** [comment]
The result looks amazing, but it will introduce a side-effect that the driver must have the cupy installed.

**Comment by rishic3** [comment]
> The result looks amazing, but it will introduce a side-effect that the driver must have the cupy installed.
> 
> Could we move the calculation of sv_normed into numpy on the driver side, and then check how much of the performance we will lose?
> 
> Or move the calculation of sv_normed based on cupy into a single spark job?

After moving sv_normed calculation into numpy on the driver side:

100,000 samples: 43s -> 49s
1,000,000 samples: 335s -> 322s (improved somehow?)

Not a big impact here at all; and in the future I can keep the driver-side free of cupy operations.

**Comment by lijinf2** [comment]
Looks good with Bobby's comment.

In addition to cupy operations, also need to move "import cupy" from driver code to worker code (i.e. make_matrix_udf). In some clusters (e.g. AWS EMR), the driver node may not be equipped with a GPU so "import cupy" may lead to a no-cuda-device error.

File: python/benchmark/gen_data_distributed.py [hunk 1]
```diff
@@ -204,26 +204,42 @@ def gen_dataframe(self, spark: SparkSession) -> Tuple[DataFrame, List[str]]:
...
-                    mode="economic",
-                    check_finite=False,
+                u, _ = cp.linalg.qr(
+                    cp.random.standard_normal(size=(n_partition_rows, n)),
+                    mode="reduced",
                 )
-                yield pd.DataFrame(data=np.dot(u, sv_normed))
+                data = cp.dot(u, cp.asarray(sv_normed))
+                del u
+                for i in range(0, n_partition_rows, maxRecordsPerBatch):
```
Comment (eordentlich) [inline python/benchmark/gen_data_distributed.py:236]: This part will be common to the different generators, I think, so for next one, would be good to find a way to share.

File: python/benchmark/gen_data_distributed.py [hunk 2]
```diff
@@ -204,26 +204,42 @@ def gen_dataframe(self, spark: SparkSession) -> Tuple[DataFrame, List[str]]:
...
+                    end_idx = (
+                        i + maxRecordsPerBatch
+                        if i + maxRecordsPerBatch < n_partition_rows
```
Comment (eordentlich) [inline python/benchmark/gen_data_distributed.py:239]: min(i + maxRecordsBatch, n_partition_rows) ?

**Comment by rishic3** [comment]
build

**Comment by eordentlich** [comment]
Can you rebase to latest branch-23.06 to avoid so many past commits being included in the PR?

**Review by wbo4958 — APPROVED** [review]
LGTM

File: python/benchmark/gen_data_distributed.py [hunk 3]
```diff
@@ -204,26 +203,38 @@ def gen_dataframe(self, spark: SparkSession) -> Tuple[DataFrame, List[str]]:
...
-                    check_finite=False,
-                )
-                yield pd.DataFrame(data=np.dot(u, sv_normed))
+                for i in range(0, n_partition_rows, maxRecordsPerBatch):
+                    end_idx = min(i + maxRecordsPerBatch, n_partition_rows)
+                    u, _ = cp.linalg.qr(
+                        cp.random.standard_normal(size=(end_idx - i, n)),
+                        mode="reduced",
+                    )
+                    data = cp.dot(u, cp.asarray(sv_normed))
```
Comment (eordentlich) [inline python/benchmark/gen_data_distributed.py:235]: cp.asarray(sv_normed) seems should be called once outside the loop

**Comment by rishic3** [comment]
Regression (low-rank case): 1,000,000 samples, 3000 features, float32, 20g memory, 75 partitions, 12 threads -> 366 seconds

**Comment by eordentlich** [comment]
build

**Review by eordentlich — APPROVED** [review]
👍

**Pull request merged**